source/NVDAObjects/behaviors.py

```diff
@@ -366,6 +367,106 @@ def event_loseFocus(self):
 		super(Terminal, self).event_loseFocus()
 		self.stopMonitoring()
 
+
+class KeyboardHandlerBasedTypedCharSupport(Terminal):
+	"""A Terminal object that also provides typed character support for
+	console applications via keyboardHandler events.
+	This approach is an alternative to monitoring the console output for
+	characters close to the caret, or injecting in-process with NVDAHelper.
+	This class relies on the toUnicodeEx Windows function, and in particular
+	the flag to preserve keyboard state available in Windows 10 1607
+	and later."""
+	#: Whether this object quickly and reliably sends textChange events
+	#: when its contents update.
+	#: Timely and reliable textChange events are required
+	#: to support password suppression.
+	_supportsTextChange = True
+	#: A queue of typed characters, to be dispatched on C{textChange}.
+	#: This queue allows NVDA to suppress typed passwords when needed.
+	_queuedChars = []
+	#: Whether the console got new text lines in its last update.
+	#: Used to determine if typed character/word buffers should be flushed.
+	_hasNewLines = False
+	#: Whether the last typed character is a tab.
+	#: If so, we should temporarily disable filtering as completions may
+	#: be short.
+	_hasTab = False
+
+	def _reportNewText(self, line):
+		# Additional typed character filtering beyond that in LiveText
+		if (
+			not self._hasTab
+			and len(line.strip()) < max(len(speech.curWordChars) + 1, 3)
+		):
+			return
+		if self._hasNewLines:
+			# Clear the typed word buffer for new text lines.
+			speech.clearTypedWordBuffer()
+			self._queuedChars = []
+		super(KeyboardHandlerBasedTypedCharSupport, self)._reportNewText(line)

```

Comment:
If you're not aware of it, this could be shortened. I don't insist on applying this here, just want to point it out:

```suggestion
		super()._reportNewText(line)
```

A whole lot easier if we ever decide to change the name of this class again 😂